utils/src/main/java/software/amazon/awssdk/utils/async/StoringSubscriber.java

```diff
@@ -0,0 +1,207 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An implementation of {@link Subscriber} that stores the events it receives for retrieval.
+ *
+ * <p>Events can be observed via {@link #peek()} and {@link #drop()}. The number of events stored is limited by the
+ * {@code maxElements} configured at construction.
+ */
+@SdkProtectedApi
+public class StoringSubscriber<T> implements Subscriber<T> {
+    /**
+     * The maximum number of elements that can be stored in this subscriber. The number of elements in {@link #elements} may be
+     * slightly higher once {@link #onComplete()} and {@link #onError(Throwable)} events are added.
+     */
+    private final int maxElements;
+
+    /**
+     * The elements stored in this subscriber. The maximum size of this queue is approximately {@link #maxElements}.
+     */
+    private final Queue<Event<T>> elements;
+
+    /**
+     * The active subscription. Set when {@link #onSubscribe(Subscription)} is invoked.
+     */
+    private Subscription subscription;
+
+    /**
+     * Whether the subscriber has been completed via {@link #onComplete()} or {@link #onError(Throwable)}.
+     */
+    private boolean done = false;
+
+    /**
+     * Create a subscriber that stores up to {@code maxElements} events for retrieval.
+     */
+    public StoringSubscriber(int maxElements) {
+        Validate.isPositive(maxElements, "Max elements must be positive.");
+        this.maxElements = maxElements;
+        this.elements = new ConcurrentLinkedQueue<>();
+    }
+
+    /**
+     * Check the first event stored in this subscriber.
+     *
+     * <p>This will return a synthetic {@code EMPTY} event if no events are currently available (outstanding demand has not yet
+     * been filled). Such events are not stored and do not need to be {@link #drop()}ped.
+     */
+    public Event<T> peek() {
+        Event<T> result = elements.peek();
+        if (result == null) {
+            return Event.empty();
+        }
+        return result;
+    }
+
+    /**
+     * Drop the first event stored in this subscriber.
+     */
+    public void drop() {
+        elements.poll();
+        subscription.request(1);
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        if (this.subscription != null) {
+            subscription.cancel();
+        }
+
+        this.subscription = subscription;
+        subscription.request(maxElements);
+    }
+
+    @Override
+    public void onNext(T t) {
+        Validate.notNull(t, "onNext(null) is not allowed.");
+
+        if (!done) {
```

Comment:
Do we need done flag and done check since per [reactive streams spec](https://github.com/reactive-streams/reactive-streams-jvm), onNext should not be invoked after onComplete or onError?

> In response to a call to Publisher.subscribe(Subscriber) the possible invocation sequences for methods on the Subscriber are given by the following protocol:
onSubscribe onNext* (onError | onComplete)?

Reply:
Yeah, I think I'm with you and should drop it. I think I was being too cautious.